rfd/0233-tsh-browser-mfa.md

```diff
@@ -0,0 +1,179 @@
+    auth ->> backend: insert /browser_authentication/<request_id>
+
+    backend ->>- auth: unblock on insert
+    auth ->> backend: upsert /browser_authentication/<request_id><br/>{public key, user, ip}
+    auth ->>+ backend: wait for state change
+
+
+    auth->>web: Browser Authentication details
+
+    Note over web: share request details with user
+    web->>auth: rpc CreateAuthenticateChallenge
+    auth->>web: MFA Challenge
+    Note over web: user auths with biometrics/passkey
+    web->>auth: rpc UpdateBrowserAuthenticationState<br/>with signed MFA challenge response
+    auth ->> backend: upsert /browser_authentication/<request_id><br/>{public key, user, ip, state=approved, mfaDevice}
+
+    backend ->>- auth: unblock on state change
+    auth->>proxy: user certificates<br/>(MFA-verified, 12 hour TTL)
+    proxy->>tsh: user certificates<br/>(MFA-verified, 12 hour TTL)
+    Note over tsh: User is logged in
+```
+
+##### Login Flow
+
+The flow can be broken down in to three sections:
+1. `tsh` initiating a browser login flow
+1. The user verifying their MFA through the browser
+1. `tsh` receiving certificates 
+
+##### `tsh` initiating a browser login flow
+
+When the user performs a `tsh login`, it will check for either an explicit
+`--mfa-mode=browser` flag or it will default to browser authentication if there
+are no other MFA methods available. The user will be prompted for their
+password. `tsh` will send a request to the proxy, which verifies the password.
+If the password is wrong, return a failure message to `tsh`. If successful, the
+proxy will send a request to the auth service and return a success message to
+`tsh` to let it know it can generate and show the browser authentication URL.
+
+The auth service will generate a Request ID that it will store on the backend
+under `/browser_authentication/<request_id>`. Using the same method as [headless
+authentication](0105-headless-authentication.md), the Request ID will be a UUID
+derived from the client's public key to prevent an attacker intercepting the
+browser login. The record will have a short TTL of 5 minutes. The auth server
+waits for a decision from the user by using a resource watcher.
+
+##### The user verifying their MFA through the browser
+
+When `tsh` generates the MFA URL, it will attempt to open the user's default
+browser. If that fails, it will print the URL for the user to click. 
+
+Once in the browser, their login session will be used to connect to the auth
+server. If the user is not already logged in, they will be prompted to do so.
+
+When authenticated, the user can view the details of the request and either
+approve or deny it. If the user approves, they will verify using their MFA
+method. If the user denies the request, it will be marked as such on the
+backend.
+
+If the request is approved, the record is approved on the backend.
+
+##### `tsh` receiving certificates
+
+If the browser authentication is successful, the auth server will unblock the
```

Comment:
If `tsh` doesn't receive credentials until after the MFA operation completes, then how does it make the request that waits for the MFA to complete?

(I assume this is already solved for headless login, but I don't recall the details. It would be good to mention them here)

Reply:
The `POST /webapi/login/browser` endpoint will be unauthenticated so that `tsh` can make a request before it has credentials. I'll put more details in here about how this is secured